### PR TITLE
RFC: dependencies/outputs: unify classes

### PR DESCRIPTION
**Pull request by efiop**
There is a lot of duplication in current Output/Dependency classes, that was
left "just in case" while developing different schemes. Since our code base is
now getting more and more stable and we now know what we need to add support
for any new dep/out type in the future, we can now safely unify most of the
code.

Signed-off-by: Ruslan Kuprieiev <ruslan@iterative.ai>

### CODE REVIEW

File: dvc/command/imp.py [hunk 1]
```diff
@@ -1,11 +1,7 @@
 import os
 
-try:
-    from urlparse import urlparse
-except ImportError:
-    from urllib.parse import urlparse
-
 import dvc.logger as logger
+from dvc.utils import urlparse
```
Comment (mroutis) [inline dvc/command/imp.py]: What I've seen before is creating a compatibility module and move all the logic that supports python 2/3 compat right there: https://github.com/requests/requests/blob/master/requests/compat.py

It is great to have it at least on `utils` :+1:

Comment (efiop) [reply dvc/command/imp.py]: That is actually much more reasonable and future-proof. I'll move it to compat.py in a sec. Thanks!

File: dvc/dependency/__init__.py [hunk 2]
```diff
@@ -39,13 +33,9 @@
...
 
-# We are skipping RemoteHTTP.PARAM_CHECKSUM because is the same as RemoteS3
-SCHEMA = {
-    DependencyBase.PARAM_PATH: str,
-    schema.Optional(RemoteLOCAL.PARAM_CHECKSUM): schema.Or(str, None),
-    schema.Optional(RemoteS3.PARAM_CHECKSUM): schema.Or(str, None),
-    schema.Optional(RemoteHDFS.PARAM_CHECKSUM): schema.Or(str, None),
-}
+SCHEMA = output.SCHEMA.copy()
+del SCHEMA[schema.Optional(OutputBase.PARAM_CACHE)]
```
Comment (mroutis) [inline dvc/dependency/__init__.py]: Also, @efiop, maybe a small comment right here explaining the difference between a dependency and an output (thus, deleting `CACHE` and `METRIC` from the schema) would be helpful for someone that's just familiarizing with the codebase.

File: dvc/output/__init__.py [hunk 3]
```diff
@@ -25,10 +27,17 @@
             's3': OutputS3,
             'gs': OutputGS,
             'ssh': OutputSSH,
-            '': OutputLOCAL}
+            'local': OutputLOCAL}
 
-SCHEMA[schema.Optional(OutputLOCAL.PARAM_CACHE)] = bool
-SCHEMA[schema.Optional(OutputLOCAL.PARAM_METRIC)] = OutputLOCAL.METRIC_SCHEMA
+# We are skipping RemoteHTTP.PARAM_CHECKSUM because is the same as RemoteS3
```
Comment (mroutis) [inline dvc/output/__init__.py]: Thinking about it, it's not that straight forward that we have three different checksums here: `md5`, `etag`, `checksum`; Maybe we could improve this comment to clarify that

File: dvc/output/base.py [hunk 4]
```diff
@@ -17,3 +20,153 @@ class OutputAlreadyTrackedError(DvcException):
...
+                 info=None,
+                 remote=None,
+                 cache=True,
+                 metric=False):
+        self.stage = stage
+        self.project = stage.project
+        self.url = path
+        self.info = info
+        self.remote = remote or self.REMOTE(self.project, {})
+        self.use_cache = False if self.IS_DEPENDENCY else cache
```
Comment (mroutis) [inline dvc/output/base.py]: This makes the distinction between deps and outs more clear :+1:

**Comment by efiop** [comment]
@mroutis Added more comments. Please take a look.

**Comment by mroutis** [comment]
Neat, @efiop , eager to merge this PR :stuck_out_tongue:

**Pull request merged**